package/network/ipv6/map/files/map.sh

```diff
@@ -140,19 +141,61 @@ proto_map_setup() {
 	      json_add_string snat_ip $(eval "echo \$RULE_${k}_IPV4ADDR")
 	    json_close_object
 	  else
-	    for portset in $(eval "echo \$RULE_${k}_PORTSETS"); do
-              for proto in icmp tcp udp; do
-	        json_add_object ""
-	          json_add_string type nat
-	          json_add_string target SNAT
-	          json_add_string family inet
-	          json_add_string proto "$proto"
-                  json_add_boolean connlimit_ports 1
-                  json_add_string snat_ip $(eval "echo \$RULE_${k}_IPV4ADDR")
-                  json_add_string snat_port "$portset"
-	        json_close_object
-              done
-	    done
+	    if [ "$snat_fix" = "1" ]; then
+			# SNAT Fix: Get all ports and full port count, excluding DONT_SNAT_TO ports
+			DONT_SNAT_TO="${dont_snat_to:-0}"
+			local portcount=0
+			local allports=""
+			for portset in $(eval "echo \$RULE_${k}_PORTSETS"); do
+				local startport=$(echo $portset | cut -d'-' -f1)
+				local endport=$(echo $portset | cut -d'-' -f2)
+				for x in $(seq $startport $endport); do
+					if ! echo "$DONT_SNAT_TO" | tr ' ' '\n' | grep -qw $x; then
+						allports="$allports $portcount : $x , "
+						portcount=$((portcount + 1))
+					fi
+				done
+			done
+			allports=${allports%??}
+			if [ -n "$FW4" ]; then
+				# SNAT Fix: Create mape table
+				if nft list tables | grep -q "table inet mape"; then
+					nft delete table inet mape
+				fi
+				nft add table inet mape
+				nft add chain inet mape srcnat { type nat hook postrouting priority 0\; policy accept\; }
+				# SNAT Fix: Create the rules to SNAT to all the ports
+				for proto in icmp tcp udp; do
+					nft add rule inet mape srcnat ip protocol $proto oifname "map-$cfg" snat ip to $(eval "echo \$RULE_${k}_IPV4ADDR") : numgen inc mod $portcount map { $allports }
+				done
+			else
+				# SNAT Fix: fw3 (untested)
+				iptables -t nat -F MAPE_CHAIN 2>/dev/null
+				iptables -t nat -X MAPE_CHAIN 2>/dev/null
+				iptables -t nat -N MAPE_CHAIN
+				iptables -t nat -A POSTROUTING -o "map-$cfg" -j MAPE_CHAIN
+				for proto in icmp tcp udp; do
+        			for port in $(echo "$allports" | sed 's/.*: //g' | tr ',' '\n' | tr -d ' '); do
```

Comment:
Do we really have to do this? Can't we store a clean array for allports and just loop through it?

Reply:
Changed to another implementation, but I'm not sure if it works properly.